ports/esp32/machine_timer.c

```diff
@@ -130,20 +133,49 @@ STATIC void machine_timer_enable(machine_timer_obj_t *self) {
 
 STATIC mp_obj_t machine_timer_init_helper(machine_timer_obj_t *self, mp_uint_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
     static const mp_arg_t allowed_args[] = {
-        { MP_QSTR_period,       MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 0xffffffff} },
         { MP_QSTR_mode,         MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 1} },
         { MP_QSTR_callback,     MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = mp_const_none} },
+#if MICROPY_PY_BUILTINS_FLOAT
+        { MP_QSTR_period,       MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = mp_const_none} },
+        { MP_QSTR_freq,         MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = mp_const_none} },
+        { MP_QSTR_period_us,    MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = mp_const_none} },
+#else
+        { MP_QSTR_period,       MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 0xffffffff} },
+        { MP_QSTR_freq,         MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 0xffffffff} },
+        { MP_QSTR_period_us,    MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 0xffffffff} },
+#endif
     };
 
     machine_timer_disable(self);
 
     mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
     mp_arg_parse_all(n_args, pos_args, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
 
-    // Timer uses an 80MHz base clock, which is divided by the divider/scalar, we then convert to ms.
-    self->period = (args[0].u_int * TIMER_BASE_CLK) / (1000 * TIMER_DIVIDER);
-    self->repeat = args[1].u_int;
-    self->callback = args[2].u_obj;
+#if MICROPY_PY_BUILTINS_FLOAT
+    if (args[2].u_obj != mp_const_none) {
+	self->period = (uint64_t) (mp_obj_get_float(args[2].u_obj) * TIMER_SCALE / 1000);
```

Comment:
It may be more accurate if the float is converted to uint64_t first, then the scaling performed....?

Reply:
Not if the period value really is a float. If I'm trying to get 600Hz by specifying a period of 1.6666ms then converting to an uint64_t first will give me 1,000Hz instead.

Reply:
Yes, sorry, you're right (I thought this was the "us" calculation).